geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseConnectionPool.scala

```diff
@@ -61,114 +58,188 @@ object HBaseConnectionPool extends LazyLogging {
     }
   )
 
-  private val connections = Caffeine.newBuilder().build(
-    new CacheLoader[(Configuration, Boolean), Connection] {
-      override def load(key: (Configuration, Boolean)): Connection = createConnection(key._1, key._2)
+  private val connections: LoadingCache[(Configuration, Boolean), CachedConnection] =  Caffeine.newBuilder().build(
+    new CacheLoader[(Configuration, Boolean), CachedConnection] {
+      override def load(key: (Configuration, Boolean)): CachedConnection = {
+        createConnection(key._1, key._2) match {
+          case SingletonConnection(connection, kerberos) => CachedConnection(connection, kerberos)
+          case c => throw new NotImplementedError(s"Expected SingletonConnection but got $c")
```

Comment:
My coffee isn't working.  What's going on here?!

createConnection can only return a SingletonConnection?  And then we turn that into a CachedConnection?

Reply:
right, the method signature for createConnection is ConnectionWrapper to hide the internal implementing classes, but it always creates a new connection. the `case c` is just in case someone changes the code, but it won't ever happen atm.

Reply:
the CachedConnection just suppresses closing the connection so it can be re-used by other data stores, vs the SingletonConnection will close the connection when disposed.